homeassistant/components/light/fibaro.py

```diff
@@ -0,0 +1,185 @@
+        value = 100
+    return max(0, min(255, ((value * 256.0) / 100.0)))
+
+
+def scaleto100(value):
+    """Scale the input value from 0-255 to 0-100."""
+    if value < 2:
+        value = 0
+    if value > 253:
+        value = 255
+    return max(0, min(100, ((value * 100.4) / 255.0)))
+
+
+def setup_platform(hass, config, add_entities, discovery_info=None):
+    """Perform the setup for Fibaro controller devices."""
+    if discovery_info is None:
+        return
+
+    add_entities(
+        [FibaroLight(device, hass.data[FIBARO_CONTROLLER])
+         for device in hass.data[FIBARO_DEVICES]['light']], True)
+
+
+class FibaroLight(FibaroDevice, Light):
+    """Representation of a Fibaro Light, including dimmable."""
+
+    def __init__(self, fibaro_device, controller):
+        """Initialize the light."""
+        self._supported_flags = 0
+        self._last_brightness = 0
+        self._color = (0, 0)
+        self._brightness = None
+        self._white = 0
+
+        self._update_lock = threading.RLock()
+        if 'levelChange' in fibaro_device.interfaces:
+            self._supported_flags |= SUPPORT_BRIGHTNESS
+        if 'color' in fibaro_device.properties:
+            self._supported_flags |= SUPPORT_COLOR
+        if 'setW' in fibaro_device.actions:
+            self._supported_flags |= SUPPORT_WHITE_VALUE
+        super().__init__(fibaro_device, controller)
+        self.entity_id = ENTITY_ID_FORMAT.format(self.ha_id)
+
+    @property
+    def brightness(self):
+        """Return the brightness of the light."""
+        return self._brightness
+
+    @property
+    def hs_color(self):
+        """Return the color of the light."""
+        return self._color
+
+    @property
+    def white_value(self):
+        """Return the white value of this light between 0..255."""
+        return self._white
+
+    @property
+    def supported_features(self):
+        """Flag supported features."""
+        return self._supported_flags
+
+    def turn_on(self, **kwargs):
+        """Turn the light on."""
+        with self._update_lock:
+            if self._supported_flags & SUPPORT_BRIGHTNESS:
+                target_brightness = None
+                if ATTR_BRIGHTNESS_PCT in kwargs:
+                    target_brightness = scaleto255(kwargs[ATTR_BRIGHTNESS_PCT])
+                elif ATTR_BRIGHTNESS in kwargs:
+                    target_brightness = kwargs[ATTR_BRIGHTNESS]
+
+                # No brightness specified, so we either restore it to
+                # last brightness or switch it on at maximum level
+                if target_brightness is None:
+                    if self._brightness < 4:
+                        if self._last_brightness:
+                            self._brightness = self._last_brightness
+                        else:
+                            self._brightness = 255
+                # We're supposed to turn it on to a very very low level,
+                # so instead, we switch it off
```

Comment:
How do we set it very very low then?

I suggest you remove all the magic constants throughout and just set it to the requested value.

Reply:
Those are UI concerns. The light platform should just do as it's told :)

(We actually changed this in the UI recently, home-assistant/home-assistant-polymer#1715, because it is frustrating to be left in the dark when trying to set a low brightness)

Reply:
I see and this is a very valid point from a system perspective. Should the remembering of previous brightness level also be done by HA? Currently I implemented it, as Fibaro does it from its UI but not from its API.